### ISSUE TITLE
Configurable product where variations are disabled on website/store view appears as salable. 

### ISSUE THREAD

**Issue by ioweb-gr** [magento/inventory#3296]
### Preconditions (*)
1. Magento 2.4.1 with MSI
2. Set Magento to hide products that are out of stock.

### Steps to reproduce (*)
1. Multiple store views
- Store1 
- Store2
- Store3
2. A configurable product with 1 variation assigned to multiple sources - assigned to a stock - assigned to Store1
- Source1 (in stock, qty 2)
- Source2 (out of stock)
- Source3 (out of stock)
3. Product is enabled in "All store views" 
4. Product is disabled in store view for Store1,Store2,Store3 
5. Visit category page which shows the product. The product appears salable but is out of stock

### Expected result (*)
1. The product does not appear

### Actual result (*)
1. The product appears in the frontend as out of stock but not possible to add to cart. 

e.g. 
in category
![image](https://user-images.githubusercontent.com/20220341/110248195-8c34b680-7f78-11eb-9e24-734416280049.png)

in product
![image](https://user-images.githubusercontent.com/20220341/110248183-8343e500-7f78-11eb-8804-a2789aae805b.png)

**Comment by ioweb-gr** [comment]
After further investigating I notice that on the stock index for the specific store view, the configurable product's total salable quantity is incorrectly counted. It appears as `instock` with salable quantity = 2 because the enabled/disabled status is not taken into account correctly. 

Could you point  me to the code responsible for building the index of the total salable quantity of the configurable products to debug it?

**Comment by ioweb-gr** [comment]
I ended up tracking it all the way to this function 

`\Magento\InventoryIndexer\Indexer\SelectBuilder::execute`

Which results in the following SQL in my case

```
SELECT `source_item`.`sku`,
       SUM(IF(source_item.status = 0, 0, quantity)) AS `quantity`,
       IF((legacy_stock_item.use_config_backorders = 0 AND legacy_stock_item.backorders <> 0 AND
           (legacy_stock_item.min_qty >= 0 OR legacy_stock_item.qty > legacy_stock_item.min_qty)) OR
          ((legacy_stock_item.use_config_manage_stock = 0 AND legacy_stock_item.manage_stock = 0)) OR
          ((legacy_stock_item.use_config_min_qty = 1 AND SUM(IF(source_item.status = 0, 0, quantity)) > 0) OR
           (legacy_stock_item.use_config_min_qty = 0 AND
            SUM(IF(source_item.status = 0, 0, quantity)) > legacy_stock_item.min_qty)) OR (product.sku IS NULL), 1,
          0)                                        AS `is_salable`
FROM `inventory_source_item` AS `source_item`
         LEFT JOIN `catalog_product_entity` AS `product` ON product.sku = source_item.sku
         LEFT JOIN `cataloginventory_stock_item` AS `legacy_stock_item`
                   ON product.entity_id = legacy_stock_item.product_id
WHERE (source_item.source_code IN
       ('glyfada', 'junior_running', 'kentriki_apothiki_melenikou', 'marousi', 'outlet_running', 'outlet_tennis'))
GROUP BY `sku`
```

Which seems to calculate the total quantity available of the product in the required sources but fails to take into account the store view and the simple product (child of configurable in this case) status in the store view. 

If the focus here is just to mark the salable quantity and then filter it out somewhere else I'm not sure where this would happen. However it seems that the isSalable function will look in the inventory_stock_X tables and if the product is salable it will appear in the lists even though it shouldn't. 

The end result on inventory_stock table is this one

![image](https://user-images.githubusercontent.com/20220341/110258542-ccf9f300-7fab-11eb-9632-01f05c77e3bc.png)

But since 

`1041A004-011-45` and `1041A004-011-46` are not enabled in the store view, the product should actually be not salable.

**Comment by ioweb-gr** [comment]
Looking up at Configurable product indexer the query is 

```
SELECT `parent_product_entity`.`sku`, SUM(stock.quantity) AS `quantity`, MAX(stock.is_salable) AS `is_salable`
FROM `inventory_stock_5` AS `stock`
         INNER JOIN `catalog_product_entity` AS `product_entity` ON product_entity.sku = stock.sku
         INNER JOIN `catalog_product_super_link` AS `parent_link` ON parent_link.product_id = product_entity.entity_id
         INNER JOIN `catalog_product_entity` AS `parent_product_entity`
                    ON parent_product_entity.entity_id = parent_link.parent_id
GROUP BY `parent_product_entity`.`sku`
```

Yielding the result as above 

![image](https://user-images.githubusercontent.com/20220341/110286599-e02db280-7fed-11eb-8379-238404b2d982.png)

and again skipping the fact that the offending products have disabled status

```
+---------------+-----+--------+
|sku            |value|store_id|
+---------------+-----+--------+
|1041A004-011-45|1    |0       |
|1041A004-011-45|2    |1       |
|1041A004-011-45|2    |4       |
|1041A004-011-45|2    |5       |
|1041A004-011-45|2    |6       |
|1041A004-011-45|2    |7       |
|1041A004-011-45|2    |8       |
|1041A004-011-45|2    |9       |
|1041A004-011-45|1    |10      |
|1041A004-011-46|1    |0       |
|1041A004-011-46|2    |1       |
|1041A004-011-46|2    |4       |
|1041A004-011-46|2    |5       |
|1041A004-011-46|2    |6       |
|1041A004-011-46|2    |7       |
|1041A004-011-46|2    |8       |
|1041A004-011-46|2    |9       |
|1041A004-011-46|1    |10      |
+---------------+-----+--------+

```

**Comment by ioweb-gr** [comment]
This seems related to  #2413

**Comment by willtran** [comment]
I don't think this issue will ever get worked on. This is because a Stock in MSI is shared between different sales channel - and product status is a website scoped attribute. I have had to bring the logic for filtering out configurable products with disabled (but saleable) children when building. the query for PLP

**Comment by karpa** [comment]
Several years later we also see this exact bug. We managed to create  a query that finds the problematic products.

```sql
SELECT
    stock.sku,
    stock.quantity,
	parent_products.quantity,
    stock.is_salable,
	parent_products.is_salable
FROM
    inventory_stock_1 AS stock
LEFT JOIN
    (
         SELECT `parent_product_entity`.`sku`, SUM(stock.quantity) AS `quantity`, MAX(stock.is_salable) AS `is_salable`
FROM `inventory_stock_1` AS `stock`
         INNER JOIN `catalog_product_entity` AS `product_entity` ON product_entity.sku = stock.sku
         INNER JOIN `catalog_product_super_link` AS `parent_link` ON parent_link.product_id = product_entity.entity_id
         INNER JOIN `catalog_product_entity` AS `parent_product_entity`
                    ON parent_product_entity.entity_id = parent_link.parent_id
GROUP BY `parent_product_entity`.`sku`
    ) AS parent_products
ON
    parent_products.sku = stock.sku
WHERE
    parent_products.is_salable = 0 AND stock.is_salable=1
```

**Comment by engcom-Delta** [comment]
Hi @ioweb-gr ,

Thanks for your reporting and collaboration.
We have tried to reproduce the issue in latest 2.4-develop instance and we are not able to reproduce the issue.Kindly refer the attached screenshot.

Steps to reproduce :

1. Multiple store views
Store1
Store2
Store3
2. A configurable product with 1 variation assigned to multiple sources - assigned to a stock - assigned to Store1
Source1 (in stock, qty 2)
Source2 (out of stock)
Source3 (out of stock)

<img width="884" height="251" alt="Image" src="https://github.com/user-attachments/assets/fd89e006-19e1-4d33-877f-6765dcb608b2" />

3. Product is enabled in "All store views"
4. Product is disabled in store view for Store1,Store2,Store3
6. Visit category page , observe product is not available in search or category page.

<img width="805" height="368" alt="Image" src="https://github.com/user-attachments/assets/4cd1ede1-3d0f-487f-a7ef-caaf565b6479" />

7. Access the product directly , observe product is OOS

<img width="748" height="356" alt="Image" src="https://github.com/user-attachments/assets/9959f1db-49ab-4e0e-bd72-7cc95aeac7d2" />

Can you please re-verify in latest 2.4-develop instance and confirm.
Thanks.

**Comment by ioweb-gr** [comment]
@engcom-Delta I'm having some issues installing MSI on the develop version 

```
Running data recurring...No linked stock found
Upgrade failed: No linked stock found
```

Which stable version should I retry to see if the steps have changed. 

Would 2.4.8-p3 be good for this test?

**Comment by engcom-Delta** [comment]
Hi @ioweb-gr ,

We have verified issue in 2.4-develop instance, Can you please verify again and confirm.

Thanks.
